lib/Parse/ParseSIL.cpp

```diff
@@ -683,6 +689,80 @@ static bool parseSILOptional(bool &Result, SILParser &SP, StringRef Expected) {
   return false;
 }
 
+namespace {
+  /// A helper class to perform lookup of IdentTypes in the
+  /// current parser scope.
+  class IdentTypeReprLookup : public ASTWalker {
+    Parser &P;
+  public:
+    IdentTypeReprLookup(Parser &P) : P(P) {}
+
+    bool walkToTypeReprPre(TypeRepr *Ty) {
+      auto *T = dyn_cast_or_null<IdentTypeRepr>(Ty);
+      auto Comp = T->getComponentRange().front();
+      if (auto Entry = P.lookupInScope(Comp->getIdentifier()))
+        if (isa<TypeDecl>(Entry)) {
+          Comp->setValue(Entry);
+          return false;
+        }
+      return true;
+    }
+  };
+} // end anonymous namespace
+
+/// Remap RequirementReps to Requirements.
+void SILParser::convertRequirements(SILFunction *F,
```

Comment:
Actually, this hack is based on what we do for SIL function types. When we parse SIL function types they call internally parseTypeIdentifier at some point, which actually does resolve the references to the generic parameters in the current generic scope of the parser. But in case of the `@_specialize` attribute, we parse the references to generic parameters before (!!!) we have parsed the generic signature. Thus, the current generic parser scope does not contain enough information at the time of parsing to resolve such references. Therefore, we need to perform this resolution after we have parsed the generic signature of the SIL function. And when we do it, we essentially mimic in IdentTypeReprLookup and convertRequirements what the SIL parser usually does.